File: golang/core/medium/my_calendar_2/my_calendar_2.go
package main

import (
	"fmt"
)

type Interval struct {
	start int
	end   int
}

type MyCalendarTwo struct {
	bookings []Interval
	overlaps []Interval
}

func Constructor() MyCalendarTwo {
	return MyCalendarTwo{
		bookings: []Interval{},
		overlaps: []Interval{},
	}
}

func (this *MyCalendarTwo) Book(start int, end int) bool {
	newInterval := Interval{start: start, end: end}

	for _, overlap := range this.overlaps {
		if overlap.start < newInterval.end && overlap.end > newInterval.start {
			return false
		}
	}

	for _, booking := range this.bookings {
		if booking.start < newInterval.end && booking.end > newInterval.start {
			this.overlaps = append(this.overlaps, Interval{
				start: max(booking.start, newInterval.start),
				end:   min(booking.end, newInterval.end),
			})
		}
	}

	this.bookings = append(this.bookings, newInterval)
	return true
}

func main() {
	myCalendar2 := Constructor()
	fmt.Println(myCalendar2.Book(10, 20))
	fmt.Println(myCalendar2.Book(15, 25))
	fmt.Println(myCalendar2.Book(20, 30))
	fmt.Println(myCalendar2.Book(5, 15))
	fmt.Println(myCalendar2.Book(8, 12))
	fmt.Printf("%v\n", myCalendar2)
}

File: typescript/data_structures/oop/my_calendar_2.ts
class MyCalendarTwo {
  private bookings: [number, number][]
  private overlaps: [number, number][] // store double booked intervals

  constructor() {
    this.bookings = []
    this.overlaps = []
  }

  book(start: number, end: number): boolean {
    // 1) check if overlap any double booking
    for (const [overlapStart, overlapEnd] of this.overlaps) {
      if (Math.max(start, overlapStart) < Math.min(end, overlapEnd)) {
          // return if overlap
          return false
      }
    }

    // 2) check where the new event overlaps with the single bookings
    for (const [bookingStart, bookingEnd] of this.bookings) {
      if (Math.max(start, bookingStart) < Math.min(end, bookingEnd)) {
        const overlapStart = Math.max(start, bookingStart)
        const overlapEnd = Math.min(end, bookingEnd)
        this.overlaps.push([overlapStart, overlapEnd])
      }
    }

    // 3) add the new booking
    this.bookings.push([start, end])
    return true
  }
}

/**
* Your MyCalendarTwo object will be instantiated and called as such:
* var obj = new MyCalendarTwo()
* var param_1 = obj.book(start,end)
*/
const myCalendarTwo = new MyCalendarTwo()
console.log(myCalendarTwo.book(10, 20))
console.log(myCalendarTwo.book(50, 60))
console.log(myCalendarTwo.book(10, 40))
console.log(myCalendarTwo.book(5, 15))
console.log(myCalendarTwo)